universal-idl-program/program/src/instructions/create_idl/mod.rs

```diff
@@ -0,0 +1,208 @@
+use crate::account_checks::assert_account_owned_by_loader;
+use crate::account_checks::assert_address;
+use crate::account_checks::assert_executable;
+use crate::account_checks::assert_mut;
+use crate::account_checks::assert_owner;
+use crate::account_checks::assert_program_authority_matches_program;
+use crate::account_checks::assert_program_data_matches_program;
+use crate::account_checks::assert_signer;
+use crate::account_checks::assert_system_program;
+use crate::account_checks::assert_with_msg;
+use crate::account_checks::is_frozen_program;
+use crate::account_checks::is_upgradeable_program;
+use crate::id;
+use crate::state::assert_frozen_authority_seeds;
+use crate::state::assert_idl_seeds;
+use crate::state::FrozenProgramAuthority;
+use crate::state::Idl;
+use crate::state::SolanaAccount;
+use crate::IdlProgramInstruction;
+
+use borsh::BorshSerialize;
+use lazy_format::lazy_format;
+use solana_program::bpf_loader_upgradeable::UpgradeableLoaderState;
+use solana_program::program::invoke_signed;
+use solana_program::{
+    account_info::{next_account_info, AccountInfo},
+    entrypoint::ProgramResult,
+    instruction::{AccountMeta, Instruction},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    system_instruction::create_account,
+    sysvar::{clock::Clock, rent::Rent, Sysvar},
+};
+
+pub fn create_idl(
+    program_id: Pubkey,
+    payer: Pubkey,
+    program_authority: Pubkey,
+    idl_account: Pubkey,
+    program: Pubkey,
+    program_data_or_frozen_auth: Pubkey,
+) -> Result<Instruction, ProgramError> {
+    Ok(Instruction {
+        program_id,
+        accounts: vec![
+            AccountMeta::new(payer, true),
+            AccountMeta::new_readonly(program_authority, true),
```

Comment:
How would this possible sign in the case that program_authority is a PDA of the current program?

Reply:
You would have to have program instructions on the PDA's program that allow CPI to this program

Reply:
I thought `program_authority` could be the `frozen_authority` derived from the current program.

Might have misread the first time around